Add loyalty card filter

TheLastProject opened the pull request:
Fixes #278

![Screenshot_1574440421](https://user-images.githubusercontent.com/1885159/69443253-543b3d00-0d4e-11ea-9a7f-4eb6dd348e36.png)
![Screenshot_1574440423](https://user-images.githubusercontent.com/1885159/69443254-543b3d00-0d4e-11ea-9c91-063a74f78fd5.png)
![Screenshot_1574452072](https://user-images.githubusercontent.com/1885159/69455803-75f5ed80-0d69-11ea-85f4-b3211ded1864.png)

TheLastProject commented:
I feel this is ready for review now :)

brarcher commented on app/src/main/java/protect/card_locker/DBHelper.java:
```diff
@@ -144,11 +144,17 @@ public boolean deleteLoyaltyCard (final int id)
     }
 
-    public Cursor getLoyaltyCardCursor()
+    public Cursor getLoyaltyCardCursor(final String filter)
```
As this call is used in two different contexts (with and without a filter), could a new get be added instead that supports the filter? e.g.

public Cursor getLoyaltyCardCursor() 
{
   // as normal
}

/**
 * Some javadoc that explains this call and what filter does
 */
public Cursor getLoyaltyCardCursor(final String filter) 
{
    // the new call that also can filter
}


Additionally, can there be a unit test that covers this new functionality? (Or are there new tests but I missed them?)

TheLastProject replied:
Oh, I forgot Java can have the same function defined multiple times with different parameters (I normally develop Python). Yeah, that would be much cleaner indeed! Will do!

I haven't added any tests, but I'll try to add some.

brarcher commented on app/src/main/java/protect/card_locker/DBHelper.java:
```diff
@@ -175,5 +175,29 @@ public int getLoyaltyCardCount()
         return numItems;
     }
+
+    public int getFilteredLoyaltyCardCount(String filter)
```
Could this call use getLoyaltyCardCursor(), so that the query is not listed twice?

TheLastProject replied:
Yup, will fix that

brarcher commented on app/src/main/res/layout/main_activity.xml:
```diff
@@ -6,7 +6,9 @@
-    tools:context="protect.card_locker.MainActivity">
+    tools:context="protect.card_locker.MainActivity"
+    android:focusable="true"
+    android:focusableInTouchMode="true">
```
Can you explain what this change is for a little?

TheLastProject replied:
I honestly don't know why this works and setting default focus on the edittext element to false doesn't, but I found this on https://android--code.blogspot.com/2015/08/android-edittext-auto-focus.html?m=1 and in my testing it worked reliably. I will add a comment to explain.

brarcher commented on app/build.gradle:
```diff
@@ -11,7 +11,7 @@ android {
     defaultConfig {
         applicationId "protect.card_locker"
-        minSdkVersion 15
+        minSdkVersion 16
```
Why is this necessary? Can you provide a reference?

TheLastProject replied:
Without this change the Travis build fails: https://travis-ci.org/brarcher/loyalty-card-locker/builds/615722153?utm_medium=notification&utm_source=github_status

brarcher commented:
Thanks for proposing some changes for this feature!

To help me in keeping this feature alive (and preventing breakage), are there new Robolectric tests which you could add to test the feature?

TheLastProject commented:
I've added a filtering test and addressed your other comments. I believe I have hereby addressed all your concerns. Please tell me if there's anything else you feel needs changing.

brarcher commented:
Tests look good, I appreciate it.

I loaded this onto a device to try it out. I've question about the search bar and when it is viewable. One pattern I've seen in other application is to have a menu item for search, that when clicked opens the search box. An example is in this tutorial:

https://youtu.be/9OWmnYPX1uc?t=57

I do not think that the search interface needs to be as complex that as that example. Maybe only hide the search bar until the search icon is clicked, then let it fill the app bar:

<img width="261" alt="Screen Shot 2019-11-23 at 9 23 43 AM" src="https://user-images.githubusercontent.com/5264535/69482656-0d435b00-0dd3-11ea-988d-028ce9250f85.png">

<img width="260" alt="Screen Shot 2019-11-23 at 9 23 54 AM" src="https://user-images.githubusercontent.com/5264535/69482658-13393c00-0dd3-11ea-9a05-21bfd3047171.png">

There is an example in another app that I wrote here:

https://github.com/brarcher/budget-watch/blob/master/app/src/main/java/protect/budgetwatch/TransactionActivity.java#L109

The relevant part of that example is setting up the search view in onCreateOptionsMenu(). The way your search filters the list on each character should be fine.

What do you think?

TheLastProject commented:
That makes sense, especially because you need to manually focus the search bar anyway. I'll take a look at that.

TheLastProject commented:
I've switched to using a SearchView, which indeed looks much better. Sadly, I can't figure out how to make the tests work at all. Got any pointers on how to send something to the search input?

brarcher commented:
I looked into how to test this a little. In the test, retrieving the menu item SearchView is straight forward:
```
final Menu menu = shadowOf(mainActivity).getOptionsMenu();
MenuItem searchItem = menu.findItem(R.id.action_search);
searchItem.expandActionView();
SearchView searchView = (SearchView) searchItem.getActionView();
```
The tricky part is getting the query text listener. There is no getOnQueryTextListener() method on SearchView. If one had the listener, one could call its methods directly and simulate text input and check that it causes the correct response, namely that the card list is properly filtered. 

To get at the listener, one approach is to make a Shadow. 

http://robolectric.org/extending/

There is no Shadow of SearchView. One could create a simple one that only intercepted the setOnQueryTextListener() call and saved the result, then provided the result in the test. 

How do you feel about creating a small Shadow for SearchView and adding it in a separate file along with the tests?

TheLastProject commented:
Sounds like a plan, I'll try my best. Quite new to Android so may take me some time

TheLastProject commented:
I made an attempt in https://github.com/brarcher/loyalty-card-locker/pull/320/commits/691497e24bc37a37916de0fd8f03e9393b4515ab but no matter how much I try I can't get it to work. It does seem to use the shadow class, but I just get a NullPointerException now during onCreateOptionsMenu :(

brarcher commented:
I gave it an attempt as well, starting from your efforts. I find that when the SearchView shadow is being used the menu item no longer has its action view. This is the reason for the NullPointerException.  Tinkering with it and looking online I could not figure out why this is the case. 

Although it does not fix this issue, I think your shadow was close. In order to send the listener to the real object, one would need a reference to the real object:
```
@RealObject private SearchView realSearchView;
``` 
then pass the listener along:
```
@Implementation
    protected void setOnQueryTextListener(SearchView.OnQueryTextListener listener) {
        onQueryTextListener = listener;
        realSearchView.setOnQueryTextListener(listener);
    }
```

You gave it a shot, and I appreciate that. Go ahead and remove the last few test commits, and I'll accept the pull request.

brarcher commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -192,21 +244,21 @@ public boolean onOptionsItemSelected(MenuItem item)
         {
             Intent i = new Intent(getApplicationContext(), LoyaltyCardEditActivity.class);
-            startActivity(i);
+            startActivityForResult(i, 1);
```
What is the significance of '1' here?

brarcher commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -179,8 +203,36 @@ public boolean onContextItemSelected(MenuItem item)
 
+        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
+        SearchView searchView = (SearchView) menu.findItem(R.id.action_search).getActionView();
+        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
```
searchManager could be null here. Could there be a null check, and if it is null the search view is not setup?

brarcher commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -38,6 +41,7 @@
 public class MainActivity extends AppCompatActivity
 {
     private static final String TAG = "LoyaltyCardLocker";
+    private Menu _menu;
```
The style in the project is to not have a leading _ in front of members. You can change this to 'menu'.

brarcher commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -47,52 +51,73 @@ protected void onCreate(Bundle savedInstanceState)
-    {
-        super.onResume();
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode == 1)
```
Oh, the request code '1' is from here. Could you have a const for '1'?
```
private static final int SEARCH_REQUEST_CODE = 1;
```

brarcher commented:
Oh, there were a few small comments about the implementation. Could you resolve those? Then I'll accept the branch.

TheLastProject commented:
I addressed your comments and I added a "filter" variable to persist the filter through resuming and so we can have some tests anyway. I have also tested manually to make sure that change didn't cause any regressions.

brarcher commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -38,7 +38,10 @@
+    private static final int MAIN_REQUEST_CODE = 1;
+
+    private Menu menu;
+    protected String filter = "";
```
It took me a bit to figure out what this was for. This is to manipulate the search query so there can still be unit tests, correct? it is a little unconventional, though it does enable unit tests so I'll be more confident the feature will keep working. Let's go with it.

brarcher replied:
Oh, you mentioned the filter variable in your comment. I looked at the code first... (:

brarcher commented:
Thanks for contributing this feature!
